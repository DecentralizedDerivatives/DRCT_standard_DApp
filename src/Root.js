import React from 'react';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import { createLogger } from 'redux-logger';
import rootReducer from './reducers';

const middlewares = [thunk];
let composeEnhancer = compose;

if (process.env.NODE_ENV === 'development') {
  middlewares.push(createLogger());
  composeEnhancer = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
}

const Root = ({ children, initialState = {} }) => {
  const store = createStore(
    rootReducer,
    initialState,
    composeEnhancer(applyMiddleware(...middlewares))
  );

  return <Provider store={store}>{children}</Provider>;
};

export default Root;
