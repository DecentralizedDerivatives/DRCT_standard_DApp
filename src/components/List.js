import React, { Component } from 'react';
import PropTypes from 'prop-types';
import TextField from './TextField';
import CircularProgress from './CircularProgress';
import Dropdown from './Dropdown';
import { Factory, token, DRCT, web3, Exchange } from '../ethereum';
import '../styles/list.css';

class List extends Component {
  static propTypes = {
    open: PropTypes.bool.isRequired,
    toggle: PropTypes.func.isRequired
  };
  constructor() {
    super();
    this.state = {
      open: false,
      selectedToken: '',
      amount: '',
      price: '',
      loading: false,
      disabled: false,
      created: false,
      myTokens: [],
      showList: false,
      txId: '',
      approval: ''
    };
  }
  componentWillMount() {
    this.getMyPositions();
  }
  handleChange = event => {
    this.setState({ [event.target.name]: event.target.value });
  };
  handleTextfieldChange = name => event => {
    this.setState({
      [name]: event.target.value
    });
  };
  getMyPositions = async () => {
    const factory = await Factory.at(
      '0x15bd4d9dd2dfc5e01801be8ed17392d8404f9642'
    );
    const accounts = await web3.eth.getAccounts();
    const numDates = await factory.getDateCount();
    let _allrows = [];
    let openDates = [];
    for (let i = 0; i < numDates; i++) {
      const startDates = (await factory.startDates.call(i)).c[0];
      const _token_addresses = await factory.getTokens(startDates);
      let _date = new Date(startDates * 1000);
      _date =
        _date.getMonth() +
        1 +
        '/' +
        _date.getDate() +
        '/' +
        _date.getFullYear();
      for (let j = 0; j < _token_addresses.length; j++) {
        const drct = await DRCT.at(_token_addresses[j]); //Getting contract
        const _balance = (await drct.balanceOf(accounts[0])).c[0]; //Getting balance of token
        if (_balance > 0) {
          _allrows.push(
            _token_addresses[j] + '(' + _balance + '/' + _date + ')'
          ); //Pushing token address + balance/date
        }
      }
    }
    console.log('mytokens', _allrows);
    _allrows.length
      ? this.setState({
          myTokens: _allrows,
          selectedToken: _allrows[0]
        })
      : this.setState({ myTokens: ['No Current Positions'] });
  };

  approveOrder = async () => {
    const accounts = await web3.eth.getAccounts();
    const exchange = await Exchange.deployed();
    const tokenSel = this.state.selectedToken
      .split('(')[0]
      .replace(/['"]+/g, '');
    const drct = await DRCT.at(tokenSel);
    let response, error;
    this.setState({ loading: true, disabled: true, showApproval: true });
    console.log('inputs', exchange.address, this.state.amount);
    try {
      response = await drct.approve(exchange.address, this.state.amount, {
        from: accounts[0],
        gas: 4000000
      });
      /*Handle Success Here*/
      this.setState({
        showList: true,
        txId: response.tx,
        approval: 'Order approval confirmed',
        loading: false
      });
    } catch (error) {
      /*Handle error here*/
      this.setState({
        txId: error.tx,
        error: true,
        disabled: false,
        loading: false,
        approval: 'Error approving order'
      });
    }
  };

  listOrder = async () => {
    const exchange = await Exchange.deployed();
    const accounts = await web3.eth.getAccounts();
    const tokenSel = this.state.selectedToken
      .split('(')[0]
      .replace(/['"]+/g, '');
    let response, error;
    exchange
      .list(tokenSel, this.state.amount, this.state.price * 1e18, {
        from: accounts[0],
        gas: 4000000
      })
      .then((res, err) => {
        if (err) {
          console.log('Error Message:', err);
        } else {
          console.log('RESPONSE', res);
        }
      });
    this.props.toggle;
  };

  render() {
    const { classes } = this.props;
    return (
      <div>
        <div className="container list-form">
          <div className="dialog-container">
            <div className="input-container">
              <p className="input">Place Order</p>
              <div className="flex-container">
                <Dropdown
                  options={this.state.myTokens}
                  value={this.state.selectedToken || 'Select a Token'}
                  name="selectedToken"
                  onChange={this.handleChange}
                  className="dropdown-selectedToken"
                />
              </div>
            </div>

            <div className="input-container">
              <p className="input">Price (in Ether)</p>

              <TextField
                id="price"
                value={Number(this.state.price)}
                type="number"
                onChange={this.handleTextfieldChange('price')}
                className="full-width"
                helperText="Enter the price in Ether (e.g. 0.1)"
              />
            </div>

            <div className="input-container">
              <p className="input">Amount</p>

              <TextField
                id="amount"
                value={Number(this.state.amount)}
                type="number"
                onChange={this.handleTextfieldChange('amount')}
                className="full-width"
                helperText="Enter the amount of the token to sell"
              />
            </div>

            <button
              className={this.state.disabled ? 'button-disabled' : 'button'}
              disabled={this.state.disabled}
              onClick={this.approveOrder}
            >
              <span className="button-text">Submit for Approval</span>
            </button>
          </div>

          {this.state.showApproval && <div className={classes.line} />}
          {this.state.showApproval && (
            <div className="approval-container">
              <div className="input-container">
                <div className="flex-container-stretch">
                  <div>
                    <p className="input">Approval</p>
                  </div>

                  <div>
                    {this.state.loading && (
                      <div className="flex-container-stretch">
                        <div>
                          <p className="waiting">Waiting for confirmation...</p>
                        </div>

                        <div>
                          <CircularProgress />
                        </div>
                      </div>
                    )}
                  </div>
                </div>

                {this.state.approval && (
                  <p className="approval input-text">{this.state.approval}</p>
                )}
              </div>
            </div>
          )}

          {this.state.showList && <div className="line" />}
          {this.state.showList && (
            <div className="list-container">
              <button className="button" onClick={this.listOrder}>
                <span className="button-text">List Order</span>
              </button>
            </div>
          )}
        </div>
      </div>
    );
  }
}

export default List;
