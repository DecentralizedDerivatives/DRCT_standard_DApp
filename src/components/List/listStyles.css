.input-container {
  marginbottom: '15px';
}

.input-title {
  font-weight: 500;
  font-size: '13px';
}

.button-text {
  color: #fff;
}

.waiting {
  font-weight: 300;
  font-size: '13px';
  color: 'rgba(0, 0, 0, 0.54)';
  position: 'relative';
  float: 'right';
}

.approval {
  color: 'green';
}

.button {
  width: 100%;
  background-color: #114770;
}

.button-disabled: {
  width: 100%;
  background-color: #9fb5c5;
}
