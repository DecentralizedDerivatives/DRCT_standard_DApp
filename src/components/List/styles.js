const width = '325px';

export default theme => ({
  paper: {
    borderRadius: '5px',
    position: 'fixed',
    top: '20%',
    outline: 'none',
  },
  dialogContent: {
    width,
    padding: '30px',
  },
  approvalContainer: {
    width,
    paddingBottom: '5px',
  },
  listContainer: {
    width,
    padding: '15px 30px 30px 30px',
  },
  line: {
    border: `0.5px solid ${theme.colors.lightGray}`,
    marginBottom: '15px',
  },
  inputContainer: {
    marginBottom: '15px',
  },
  title: {
    fontWeight: theme.fonts.weight.bold,
    fontSize: theme.fonts.size.inputTitle,
  },
  duration: {
    width: '150px',
  },
  currency: {
    width: '150px',
  },
  fullWidth: {
    width: '100%',
  },
  txId: {
    fontSize: '16px',
  },
  button: {
    width: '100%',
    backgroundColor: theme.colors.blue,
    '&:hover': {
      backgroundColor: theme.colors.lightBlue,
    },
  },
  buttonDisabled: {
    width: '100%',
    backgroundColor: theme.colors.lightBlue,
  },
  buttonText: {
    color: theme.colors.white,
  },
  waiting: {
    fontWeight: theme.fonts.weight.light,
    fontSize: theme.fonts.size.inputTitle,
    color: theme.colors.gray,
    position: 'relative',
    float: 'right',
  },
  progress: {
    marginLeft: '10px',
    color: theme.colors.gray,
  },
});
