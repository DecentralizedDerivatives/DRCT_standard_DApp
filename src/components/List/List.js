import React, {Component} from 'react';
import PropTypes from 'prop-types';
import withStyles from 'material-ui/styles/withStyles';
import Button from 'material-ui/Button';
import TextField from 'material-ui/TextField';
import Grid from 'material-ui/Grid';
import Typography from 'material-ui/Typography';
import Dialog, {DialogContent} from 'material-ui/Dialog';
import {DatePicker} from 'material-ui-pickers';
import {CircularProgress} from 'material-ui/Progress';
import styles from './styles';
import Dropdown from '../Dropdown';
import {Factory, token,DRCT, web3, Exchange} from '../../ethereum';

class List extends Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    open: PropTypes.bool.isRequired,
    toggle: PropTypes.func.isRequired,
  };

  state = {
    open: false,
    selectedToken: '',
    amount:"",
    price:"",
    loading: false,
    disabled: false,
    created: false,
    myTokens:[]
  };

    componentDidMount() {
      this.getMyPositions();
    }


  handleChange = event => {
    this.setState({[event.target.name]: event.target.value});
  };

  handleTextfieldChange = name => event => {
    this.setState({
      [name]: event.target.value,
    });
  };

  getMyPositions = async () =>{
    const factory = await Factory.deployed();
    const accounts = await web3.eth.getAccounts();
    let _row;
    var _allrows = [];
    var openDates = [];
    const numDates = await factory.getDateCount();
    console.log('numDates',numDates);
          for(let i=0;i<numDates;i++){
              let _date = await factory.startDates.call(i);
              _date = _date.c[0]; 
              let _token_addresses = await factory.getTokens(_date);
                var _date = new Date(_date*1000);
                var _date = (_date.getMonth() + 1) + '/' + _date.getDate() + '/' + _date.getFullYear() 
              for(let j=0;j<2;j++){
                  let drct = await DRCT.at(_token_addresses[j]);
                  let _balance = await drct.balanceOf(accounts[0]);
                  if(_balance.c[0]>0){
                    _row = _token_addresses[j] + '('+_balance.c[0].toString()+'/'+_date.toString()+')';
                    _allrows.push(_row)
                    this.setState({myTokens: _allrows});
                    if(_allrows.length == 1){
                        this.setState({selectedToken:_token_addresses[j]});
                    }
                  }
              }
       }
        if(this.state.myTokens.length == 0){
              this.setState({myTokens: ["No Current Positions"]});
        }

  }

  listOrder= async () => {
    const exchange = await Exchange.deployed();
    const accounts = await web3.eth.getAccounts();
    var string = this.state.selectedToken;
    var tokenSel = string.split('(');


    let response, error;
    console.log(this.state.price);
    console.log('INPUTS',tokenSel[0].replace(/['"]+/g, ''),this.state.amount,this.state.price*1e18);
    try {
      response = await exchange.list(tokenSel[0].replace(/['"]+/g, ''),this.state.amount,this.state.price*1e18,{
        from: accounts[0],
        gas: 4000000,
      });
    } catch (err) {
      error = err;
    }
    if (error) {
      console.log(error);
    }
  };
  render() {
    const {classes} = this.props;

    return (
      <div>
        <Dialog
          open={this.props.open}
          onClose={this.props.toggle}
          PaperProps={{className: classes.paper}}
        >
          <DialogContent className={classes.dialogContent}>
            <div className={classes.inputContainer}>
              <Typography className={classes.title}>Place Order</Typography>
                <Grid item>
                  <Dropdown
                    menuItems={this.state.myTokens}
                    value={this.state.selectedToken}
                    name="selectedToken"
                    onChange={this.handleChange}
                    className={classes.selectedToken}
                  />
                </Grid>
            </div>

            <div className={classes.inputContainer}>
              <Typography className={classes.title}>Price (in Ether)</Typography>

              <TextField
                id="price"
                value={Number(this.state.price)}
                type="number"
                onChange={this.handleTextfieldChange('price')}
                className={classes.fullWidth}
                helperText="Enter the price in Ether (e.g. 0.1)"
              />
            </div>


            <div className={classes.inputContainer}>
              <Typography className={classes.title}>Amount</Typography>

              <TextField
                id="amount"
                value={Number(this.state.amount)}
                type="number"
                onChange={this.handleTextfieldChange('amount')}
                className={classes.fullWidth}
                helperText="Enter the amount of the token to sell"
              />
            </div>

            <Button
              className={
                this.state.disabled ? classes.buttonDisabled : classes.button
              }
              disabled={this.state.disabled}
              onClick={this.listOrder}
            >
              <Typography className={classes.buttonText}>
                Submit
              </Typography>
            </Button>
          </DialogContent>
        </Dialog>
      </div>
    );
  }
}

export default withStyles(styles)(List);
