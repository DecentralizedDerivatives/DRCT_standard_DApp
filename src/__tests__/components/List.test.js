import React from 'react';
import { shallow } from 'enzyme';
import Root from '../../Root';

import List from '../../components/List';

describe('<List />', () => {
  describe('render()', () => {
    it('renders the component', () => {
      const wrapper = shallow(
        <Root>
          <List />
        </Root>
      );
      expect(wrapper).toMatchSnapshot();
    });
  });
});
