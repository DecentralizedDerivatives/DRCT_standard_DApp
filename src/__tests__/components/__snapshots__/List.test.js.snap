// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List /> render() renders the component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <List
    listOrderApproved={false}
    listOrderError={null}
    listOrderId=""
    refreshPage={[MockFunction]}
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "unsubscribe": [Function],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="order-btn"
          onClick={[Function]}
        >
          <label
            className="order-btn-label"
          >
            List Order
          </label>
        </div>,
        null,
      ],
      "className": "order-btn-wrapper",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <label
            className="order-btn-label"
          >
            List Order
          </label>,
          "className": "order-btn",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "List Order",
            "className": "order-btn-label",
          },
          "ref": null,
          "rendered": "List Order",
          "type": "label",
        },
        "type": "div",
      },
      null,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="order-btn"
            onClick={[Function]}
          >
            <label
              className="order-btn-label"
            >
              List Order
            </label>
          </div>,
          null,
        ],
        "className": "order-btn-wrapper",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <label
              className="order-btn-label"
            >
              List Order
            </label>,
            "className": "order-btn",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "List Order",
              "className": "order-btn-label",
            },
            "ref": null,
            "rendered": "List Order",
            "type": "label",
          },
          "type": "div",
        },
        null,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
